### PR TITLE
compartment-mapper: add support for package.json exotic dependency fields

### LINKED ISSUE #1177: Compartment mapper support for peerDependencies
Compartment mapper does not yet provide any behavior for `peerDependencies`. If I understand the purpose of that feature correctly, the compartment mapper should treat these the same way as `dependencies` and that the distinction between the two kinds of dependencies is otherwise only meaningful to the package manager.

If you can understand https://nodejs.org/es/blog/npm/peer-dependencies/, you’re smarter than me.

### PR DESCRIPTION
**Pull request by kumavis**
Fixes #1177 and related compatibility issues

### CODE REVIEW

File: packages/compartment-mapper/src/node-modules.js [hunk 1]
```diff
@@ -252,24 +252,49 @@ const graphPackage = async (
...
+    devDependencies = {},
+  } = packageDescriptor;
+  const allDependencies = {};
+  assign(allDependencies, dependencies);
+  assign(allDependencies, peerDependencies);
+  Object.keys(peerDependenciesMeta).forEach(name => {
+    if (peerDependenciesMeta[name].optional) {
+      optionals.add(name);
+    }
+  });
```
Comment (kriskowal) [inline packages/compartment-mapper/src/node-modules.js:273]: Please consider, in keeping with the local rustic æsthetic:

```suggestion
  for (const [name, { optional }] of Object.entries(peerDependenciesMeta)) {
    if (optional) {
      optionals.add(name);
    }
  }
```

If we went more toward the array functional style, I’d be suggesting `entries`, `filter` and `forEach(set.add, set)`, but my prevailing preference is to use loops.

The contents of `package.json` are user data to us. We probably should be validating the package descriptor with something similar to what we use to validate compartment maps, before we even see it here. I don’t think we’re validating package descriptors at all.

Filed https://github.com/endojs/endo/issues/1357

File: packages/compartment-mapper/src/node-modules.js [hunk 2]
```diff
@@ -252,24 +252,49 @@ const graphPackage = async (
   graph[packageLocation] = /** @type {Node} */ (result);
 
   /** @type {Record<string, string>} */
-  const dependencies = {};
+  const dependenciesObj = {};
```
Comment (kriskowal) [inline packages/compartment-mapper/src/node-modules.js:255]: Please consider a meaningful distinction in the name.
```suggestion
  const dependencyLocations = {};
```

File: packages/compartment-mapper/test/fixtures-optional/node_modules/optional-cjs/index.js [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  } catch (err) {
+    results.errors.one = err;
+  }
+  try {
+    results.exports.two = require('missing-two');
+  } catch (err) {
+    results.errors.two = err;
+  }
+  return results
+}
```
Comment (kriskowal) [inline packages/compartment-mapper/test/fixtures-optional/node_modules/optional-cjs/index.js:24]: Such Windows file without newline on the end.

**Review by kriskowal — APPROVED** [review]
Yeah, that ought to do the trick.

File: packages/compartment-mapper/test/fixtures-optional/node_modules/optional-cjs/index.js (hunk 3 above)
Comment (kumavis) [reply packages/compartment-mapper/test/fixtures-optional/node_modules/optional-cjs/index.js:24]: tell it to the linter rules :tongue:

**Pull request merged**